File: leetcode-everyday/动态规划/子序列问题/300-最长递增子序列/300-最长递增子序列.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var lengthOfLIS = function(nums) {
  let n = nums.length;
  if(n <= 1) return n;
  const dp = [];
  dp.push(nums[0]);    
  for(let i = 1; i < n; ++i) {
      if(dp[dp.length-1] < nums[i]) {
          dp.push(nums[i]);
      } else {
          let index = lower_bound(dp, nums[i]);
          dp[index] = nums[i];
      }
  }    
  return dp.length;
};

function lower_bound(dp, target) {
  let left = 0,
      right = dp.length-1;
  while(left <= right) {
      let mid = left + Math.floor((right - left) / 2);
      if(dp[mid] < target) {
          left = mid + 1; 
      } else {
          right = mid - 1;
      }
  }
  return left;
}

File: leetcode-everyday/动态规划/子序列问题/300-最长递增子序列/300-最长递增子序列.cpp
class Solution {
public:    
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        if (n <= 1) return n;
        vector<int> dp(n, 1);         
        int ans = 1;
        for(int i = 0; i < n; ++i) {            
            for(int j = 0; j < i; ++j) {
                if(nums[j] < nums[i]) {
                    dp[i] = max(dp[i], dp[j] + 1);                    
                }
            }
            ans = max(dp[i], ans);
        }        
        return ans;
    }
};

class Solution {
public:    
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        if (n <= 1) return n;
        vector<int> dp;
        dp.push_back(nums[0]);
        for(int i = 1; i < n; ++i) {
            if(dp.back() < nums[i]) {
                dp.push_back(nums[i]);
            } else {
                auto itr = lower_bound(dp.begin(), dp.end(), nums[i]);                
                *itr = nums[i];                
            }
        }
        return dp.size();
    }
};